Initial Dapr Shared Docs (#4121)
dapr/docs: daprdocs/content/en/operations/dapr-shared/dapr-shared.md

---
type: docs

> title: "Different Deployment Strategies (Dapr Shared)"

Reviewer: ```suggestion
title: "Dapr Shared"
```

linkTitle: "Dapr Shared"
weight: 20000

> description: "Learn more about using other deployment strategies for Dapr"

Reviewer: ```suggestion
description: "Learn more about using Dapr Shared as an alternative deployment strategy"
```

---

> By default, Dapr automatically inject a sidecar to enable the Dapr APIs for your applications, but sometimes this approach (using sidecars) might not be optimal for your use cases. 

Reviewer: ```suggestion
By default, Dapr automatically injects a sidecar to enable the Dapr APIs for your applications. However, using sidecars might not be optimal for your specific use case. 
```

> Dapr Shared enable differenet deployment strategies to create Dapr Applications using Kubernetes `Daemonset` or `Deployment`. 

Reviewer: ```suggestion
Dapr Shared enables different deployment strategies to create Dapr applications using Kubernetes `Daemonset` or `Deployment`. 
```

> By running daprd as a Kubernetes `DaemonSet` resource, the daprd container will be running in each Kubernetes Node, reducing the network hops between the applications and Dapr. You can also choose to run Dapr Shared as a Kubernetes `Deployment`, in which case, the Kubernetes scheduler will decide in which node the Dapr Shared instance will run.

Reviewer: ```suggestion
- **`DaemonSet`:** When running daprd as a Kubernetes `DaemonSet` resource, the daprd container runs in each Kubernetes node, reducing network hops between the applications and Dapr. 
- **`Deployment`:** When running Dapr Shared as a Kubernetes `Deployment`, the Kubernetes scheduler decides in which node the Dapr Shared instance will run.
```

{{% alert title="Dapr Shared deployments" color="primary" %}}

> For each Dapr Application, you need to deploy the Dapr Shared chart using different `shared.appId`s.

Reviewer: ```suggestion
For each Dapr application, you need to deploy the Dapr Shared chart using different `shared.appId`s.
```

{{% /alert %}}

## Why Dapr Shared?

> By default, when Dapr is installed into a Kubernetes Cluster, the Dapr Control Plane is in charge of injecting the daprd sidecar to workloads annotated with Dapr annotations (`dapr.io/enabled: "true"`). This mechanism delegates the responsibility of defining which workloads will be interacting with the Dapr APIs to the team in charge of deploying and configuring these workloads. Sidecars had the advantage of being co-located with your applications, so all communication between the application and the sidecar happens without involving the network.

Reviewer: ```suggestion
By default, when Dapr is installed into a Kubernetes cluster, the Dapr control plane injects a Dapr as sidecar to applications annotated with Dapr annotations ( `dapr.io/enabled: "true"`). This mechanism delegates the responsibility of defining which workloads will be interacting with the Dapr APIs to the team in charge of deploying and configuring these workloads. Sidecars have the advantage of existing alongside your applications, so all communication between the application and the sidecar happens without involving the network.
```

I'm having a hard time understanding this sentence - is there a better way to word this?
> This mechanism delegates the responsibility of defining which workloads will be interacting with the Dapr APIs to the team in charge of deploying and configuring these workloads.

Reviewer 2: See my comment below as a preferred alternative to this comment from Hannah

<img src="/images/dapr-shared/sidecar.png" width=800 style="padding-bottom:15px;">

> While sidecars are the default strategy, there are some use cases that require other approaches. For example, you want to decouple the lifecycle of your workloads from the Dapr APIs. A typical example of this is Functions, or function as a service runtimes, which might automatically downscale your idle workloads to free up resources. For such cases, keeping the Dapr APIs and all the Dapr async functionalities (such as subscriptions) might be required. Dapr Shared was created exactly for this kind of scenario.

Reviewer: ```suggestion
While sidecars are Dapr's default deployment strategy, some use cases require other approaches. Let's say you want to decouple the lifecycle of your workloads from the Dapr APIs. A typical example of this is functions, or function-as-a-service runtimes, which might automatically downscale your idle workloads to free up resources. For such cases, keeping the Dapr APIs and all the Dapr async functionalities (such as subscriptions) separate might be required. 
```

> Dapr Shared extends the Dapr sidecar model with two new deployment strategies: DaemonSet and Deployment.

Reviewer: ```suggestion
Dapr Shared was created exactly for this kind of scenario, extending the Dapr sidecar model with two new deployment strategies: `DaemonSet` and `Deployment`.
```

> No matter which strategy you choose, it is important to understand that in most use cases you will have one instance of Dapr Shared (Helm Release) per service (app-id). This means that if you have an application composed by three microservices, each service is recommended to have it's own Dapr Shared instance. Check the step-by-step tutorial using Kubernetes KinD here, to see an application using Dapr Shared.

Reviewer: ```suggestion
{{% alert title="Important" color="primary" %}}
No matter which strategy you choose, it is important to understand that in most use cases, you will have one instance of Dapr Shared (Helm release) per service (app-id). This means that if you have an application composed of three microservices, each service is recommended to have its own Dapr Shared instance. Check the step-by-step tutorial using Kubernetes KinD to see an application using Dapr Shared.
{{% /alert %}}
```

It looks like you need to add a link here - does this tutorial currently exist?
> Check the step-by-step tutorial using Kubernetes KinD to see an application using Dapr Shared.

> ### Dapr Shared: DeamonSet strategy

Reviewer: ```suggestion
### `DeamonSet` strategy
```

> Kubernetes `DaemonSets` allows you to define workloads that need to be deployed once per node in the cluster. This enables workloads that are running in the same node to communicate with local daprd APIs, no matter where the `Kubernetes Scheduler` schedules your workload.

Reviewer: ```suggestion
With Kubernetes `DaemonSet`, you can define workloads that need to be deployed once per node in the cluster. This enables workloads that are running in the same node to communicate with local daprd APIs, no matter where the Kubernetes `Scheduler` schedules your workload.
```

<img src="/images/dapr-shared/daemonset.png" width=800 style="padding-bottom:15px;">

> {{% alert title="Dapr Shared DaemonSet resources" color="primary" %}}

Reviewer: ```suggestion
{{% alert title="Note" color="primary" %}}
```

> Note that DaemonSet, because it installs one instance per node, will consume more overall resources in your cluster.

Reviewer: ```suggestion
Since `DaemonSet` installs one instance per node, it will consume more overall resources in your cluster.
```

{{% /alert %}}

> ### Dapr Shared: Deployment strategy

Reviewer: ```suggestion
### `Deployment` strategy
```

> Kubernetes Deployments in the other hand, are installed once per cluster and the Kubernetes Scheduler decides, based on available resources, in which node the workload will be scheduled. For Dapr Shared, this means that your workload and the daprd instance might be located in separate nodes, which can introduce considerable network latency.

Reviewer: ```suggestion
Kubernetes `Deployments` are installed once per cluster. Based on available resources, the Kubernetes `Scheduler` decides in which node the workload will be scheduled. For Dapr Shared, this means that your workload and the daprd instance might be located in separate nodes, which can introduce considerable network latency.
```

<img src="/images/dapr-shared/deployment.png" width=800 style="padding-bottom:15px;">

## Getting Started with Dapr Shared

> {{% alert title="Install Dapr before using Dapr Shared" color="primary" %}}

Reviewer: ```suggestion
{{% alert title="Prerequisites" color="primary" %}}
```

> Before installing Dapr Shared, please ensure you have Dapr installed in your cluster.

Reviewer: ```suggestion
Before installing Dapr Shared, make sure you have [Dapr installed in your cluster]({{< ref "kubernetes-deploy.md" >}}).
```

{{% /alert %}}

If you want to get started with Dapr Shared, you can easily create a new Dapr Shared instance by installing the official Helm Chart:

```
helm install my-shared-instance oci://registry-1.docker.io/daprio/dapr-shared-chart --set shared.appId=<DAPR_APP_ID> --set shared.remoteURL=<REMOTE_URL> --set shared.remotePort=<REMOTE_PORT>
```

> 

Reviewer: This doc will need "Next steps" or "Related links" section to cross-link with relevant docs or sample tutorials... what links do you think would work best here?

```suggestion
## Next steps
```

---
type: docs

> title: "Dapr Shared"

Reviewer 2: ```suggestion
title: "Deploy Dapr per-node or per-cluster with Dapr Shared"
```

Author: I would use per-node. Machine is not a Kubernetes concept

Reviewer 2: Yes, ok, changed suggestion

linkTitle: "Dapr Shared"
weight: 20000

> description: "Learn more about using Dapr Shared as an alternative deployment strategy"

Reviewer 2: ```suggestion
description: "Learn more about using Dapr Shared as an alternative deployment to sidecars"
```

---

> By default, Dapr automatically injects a sidecar to enable the Dapr APIs for your applications. However, using sidecars might not be optimal for your specific use case. 

Reviewer 2: This section is more of the "how" and should be moved after the "why" section below.

> By default, Dapr automatically injects a sidecar to enable the Dapr APIs for your applications. However, using sidecars might not be optimal for your specific use case. 

Reviewer 2: ```suggestion
Dapr automatically injects a sidecar to enable the Dapr APIs for your applications for the best availability and reliability. 
```

> Dapr Shared enables different deployment strategies to create Dapr applications using Kubernetes `Daemonset` or `Deployment`. 

Reviewer 2: ```suggestion
Dapr Shared enables two alternative deployment strategies to create Dapr applications using a Kubernetes `Daemonset`for a per-node deployment or a`Deployment` for a per-cluster deployment. 
```

> - **`DaemonSet`:** When running daprd as a Kubernetes `DaemonSet` resource, the daprd container runs in each Kubernetes node, reducing network hops between the applications and Dapr. 

Reviewer 2: ```suggestion
- **`DaemonSet`:** When specifying to run Dapr as a Kubernetes `DaemonSet` resource, the daprd container runs on each Kubernetes node. This can reduce network hops between the applications and Dapr. 
```

- **`Deployment`:** When running Dapr Shared as a Kubernetes `Deployment`, the Kubernetes scheduler decides in which node the Dapr Shared instance will run.

{{% alert title="Dapr Shared deployments" color="primary" %}}

> For each Dapr application, you need to deploy the Dapr Shared chart using different `shared.appId`s.

Reviewer 2: I do not understand this comment. Please explain more.

Author: @msfussell I wonder how can we make it more clear, but this refers to the fact that what you are deploying at the end of the day is just `daprd`, so if you have multiple applications you will need separate Dapr Shared "instances" to make sure that we don't break app identity. So if you have 3 different applications that need to use the Dapr APIs, you will need to deploy Dapr Shared 3 times, each with its own `shared.appId` value (this is the name of the app that you want to use for a specific instance)

Reviewer 2: Doesn't that defeat the purpose of Dapr shared, that there is one per node? This goes to my comment at the end of having a working example of Dapr Shared based on the Hello World Kubernetes Tutorial. That way you can see this in action.

{{% /alert %}}

> ## Why Dapr Shared?

Reviewer 2: Suggest moving this heading to be first to the "why" of Dapr shared is explained first.

> By default, when Dapr is installed into a Kubernetes Cluster, the Dapr Control Plane is in charge of injecting the daprd sidecar to workloads annotated with Dapr annotations (`dapr.io/enabled: "true"`). This mechanism delegates the responsibility of defining which workloads will be interacting with the Dapr APIs to the team in charge of deploying and configuring these workloads. Sidecars had the advantage of being co-located with your applications, so all communication between the application and the sidecar happens without involving the network.

Reviewer 2: ```suggestion
By default, when Dapr is installed into a Kubernetes cluster, the Dapr control plane injects a Dapr as sidecar to applications annotated with Dapr annotations ( `dapr.io/enabled: "true"`). Sidecars have many advantages including improved resiliency since there is an instance per application and all communication between the application and the sidecar happens without involving the network.
```

<img src="/images/dapr-shared/sidecar.png" width=800 style="padding-bottom:15px;">

> While sidecars are Dapr's default deployment strategy, some use cases require other approaches. Let's say you want to decouple the lifecycle of your workloads from the Dapr APIs. A typical example of this is functions, or function-as-a-service runtimes, which might automatically downscale your idle workloads to free up resources. For such cases, keeping the Dapr APIs and all the Dapr async functionalities (such as subscriptions) separate might be required. 

Reviewer 2: ```suggestion
While sidecars are Dapr's default and highly recommended deployment strategy, there are some use cases require other approaches. 

For example one scenario may be to reduce the resource usage on your cluster either for local  development or or testing purpose. In this case rather than running Dapr on a per instance basis, using Dapr Share you can have Dapr per cluster. Note that although this saves resources, this signifcantly reduces the reliability of your applications since there is a single point of failure and it is not recommended to use Dapr Shared in production Kubernetes environments. There may be cases such as constrained edge devices where Dapr Shared is the best choice. Another example of this is calling the Dapr APIs over HTTP from a WASM module where Dapr is unable to be deployed. 

Another scenario is that you want to decouple the lifecycle of your workloads from the Dapr. A example of this is functions, or function-as-a-service runtimes, which might automatically downscale your idle functions to free up resources. For such cases, keeping the Dapr and all the Dapr async functionalities (such as subscriptions) separate might be required. 
```

Author: @msfussell there are several comments that might indicate that you can save resources using Dapr Shared.. and I don't think that is true. The main advantage of using Dapr Shared is to keep `daprd` running while your application might be downscaled. With Dapr Shared the application lifecycle and the `daprd` lifecycle are separated. I don't think it has anything to do with resource usage. 

Also this comment indicates that Dapr Shared is not recommended for production, which I don't fully understand. Remember that Dapr Shared is not a multi tenant `daprd` it is just `daprd` that doesn't run as a sidecar. So I don't agree with the first half of this suggestion. 

```
Another scenario is that you want to decouple the lifecycle of your workloads from the Dapr. A example of this is functions, or function-as-a-service runtimes, which might automatically downscale your idle functions to free up resources. For such cases, keeping the Dapr and all the Dapr async functionalities (such as subscriptions) separate might be required. 
```

That second part is accuarate

> Dapr Shared was created exactly for this kind of scenario, extending the Dapr sidecar model with two new deployment strategies: `DaemonSet` and `Deployment`.

Reviewer 2: ```suggestion
Dapr Shared was created for these scenario, extending the Dapr sidecar model with two new deployment approaches: `DaemonSet` and `Deployment`.
```

> No matter which strategy you choose, it is important to understand that in most use cases you will have one instance of Dapr Shared (Helm Release) per service (app-id). This means that if you have an application composed by three microservices, each service is recommended to have it's own Dapr Shared instance. Check the step-by-step tutorial using Kubernetes KinD here, to see an application using Dapr Shared.

Reviewer 2: ```suggestion
No matter which approach you choose, it is important to understand that in most use cases you have one instance of Dapr Shared (Helm Release) per service (app-id). This means that if you have an application composed by three microservices, each service is recommended to have it's own Dapr Shared instance. Check the step-by-step tutorial using Kubernetes Kind here, to see an application using Dapr Shared.
```

Reviewer 2: Turn this "tep-by-step tutorial using Kubernetes Kind here" into a link

### `DeamonSet` strategy

> With Kubernetes `DaemonSet`, you can define workloads that need to be deployed once per node in the cluster. This enables workloads that are running in the same node to communicate with local daprd APIs, no matter where the Kubernetes `Scheduler` schedules your workload.

Reviewer 2: ```suggestion
With Kubernetes `DaemonSet`, you can define workloads that need to be deployed once per node in the cluster. This enables workloads that are running on the same node to communicate with local Dapr APIs, no matter where the Kubernetes `Scheduler` schedules your workload.
```

<img src="/images/dapr-shared/daemonset.png" width=800 style="padding-bottom:15px;">

{{% alert title="Note" color="primary" %}}

> Since `DaemonSet` installs one instance per node, it will consume more overall resources in your cluster.

Reviewer 2: What does this comment mean? You mean more resource than per cluster, but this is less than as a side-car. I find this comment confusing. May be best to remove altogether.

Author: When you use DaemonSet mode, your `daprd` for each Dapr application will be scheduled in each node. So if you have a 3 cluster node you will have three instances. I think this explanation needs to be here, as it is a very important aspect of why we have this mode.

{{% /alert %}}

### `Deployment` strategy

> Kubernetes `Deployments` are installed once per cluster. Based on available resources, the Kubernetes `Scheduler` decides in which node the workload will be scheduled. For Dapr Shared, this means that your workload and the daprd instance might be located in separate nodes, which can introduce considerable network latency.

Reviewer 2: ```suggestion
Kubernetes `Deployments` are installed once per cluster. Based on available resources, the Kubernetes `Scheduler` decides in which node the workload is scheduled. For Dapr Shared, this means that your workload and the Dapr instance might be located on separate nodes, which can introduce considerable network latency with the trade of of reduce resource usage.
```

<img src="/images/dapr-shared/deployment.png" width=800 style="padding-bottom:15px;">

## Getting Started with Dapr Shared

{{% alert title="Prerequisites" color="primary" %}}

> Before installing Dapr Shared, make sure you have [Dapr installed in your cluster]({{< ref "kubernetes-deploy.md" >}}).

Reviewer 2: ```suggestion
Before installing Dapr Shared, make ensure you have [Dapr installed in your cluster]({{< ref "kubernetes-deploy.md" >}}).
```

{{% /alert %}}

> If you want to get started with Dapr Shared, you can easily create a new Dapr Shared instance by installing the official Helm Chart:

Reviewer 2: ```suggestion
If you want to get started with Dapr Shared, you can create a new Dapr Shared instance by installing the official Helm Chart:
```

```
helm install my-shared-instance oci://registry-1.docker.io/daprio/dapr-shared-chart --set shared.appId=<DAPR_APP_ID> --set shared.remoteURL=<REMOTE_URL> --set shared.remotePort=<REMOTE_PORT>

> ```

Reviewer 2: Surely you do more steps that just this Helm chart. I have to add an annotation correct? Can you provide 1,2,3... steps. How do I try this out for both the deployment modes.

Reviewer 2: The best approach to me, is that in the Dapr Shared repo you have another example tutorial that uses the https://github.com/dapr/quickstarts/tree/master/tutorials/hello-kubernetes tutorial

Author: no annotations is needed.

> ## Next steps

Reviewer 2: Add 
- Link to the sample repo that you had above.

Reviewer 2: @hhunter-ms - Can you add the updated diagrams here?

Reviewer: @msfussell - I sent the diagrams to @salaboy!
